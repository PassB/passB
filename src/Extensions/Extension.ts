export interface ActionMetaData {
    [key: string]: string;
}

export interface ListEntry {
    label: string;
    actions: string[];
    metadata: ActionMetaData;
}

export interface Options {
}

export type RegisterEntryCallback = (entry: ListEntry) => void;

export abstract class Extension {
    public abstract readonly name: string;
    public abstract readonly actions: string[];
    protected readonly options: Options;

    public constructor(options: Options) {
        this.options = options;
    }

    public abstract initializeList(registerEntryCallback: RegisterEntryCallback): void;

    public abstract executeAction(action: string, entry: ListEntry): void;
}
