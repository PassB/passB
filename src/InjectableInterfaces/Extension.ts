import {RouteProps} from 'react-router';
import {ExtensionName} from 'State/Interfaces';
import {MapTypeAllowedData, TypedMap} from 'State/Types/TypedMap';
import {OptionsPanelType} from './OptionsPanel';

export interface Extension<OptionType extends MapTypeAllowedData<OptionType>> {
  readonly routes: RouteProps[];
  readonly OptionsPanel?: OptionsPanelType<OptionType>;
  readonly actions: string[];
  readonly name: ExtensionName;
  readonly defaultOptions: TypedMap<OptionType>;
  initializeList(): Promise<void>;
  getLabelForAction(action: string): string;
  executeAction(action: string, entry: string, options: ExecutionOptions): void;

}

export interface ExecutionOptions {
  navigateTo: (newUrl: string, state: {}) => void;
}
