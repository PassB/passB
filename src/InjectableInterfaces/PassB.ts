import {Extension, FileFormat, Filler, Matcher} from '.';

export interface PassB {
  getAllExtensions(): Array<Extension<{}>>;

  getExtension(name: string): Extension<{}>;

  getMatcher(): Matcher<{}>;

  getAllMatchers(): Array<Matcher<{}>>;

  getFiller(): Filler<{}> ;

  getAllFillers(): Array<Filler<{}>> ;

  getFileFormat(): FileFormat<{}>;

  getAllFileFormats(): Array<FileFormat<{}>> ;

  reloadEntries(): Promise<this> ;

  reloadExtension(): Promise<{}>;
}
