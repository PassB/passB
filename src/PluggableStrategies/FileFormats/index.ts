export {FileFormat, FileFormatTag} from './FileFormat';

import './FirstLineFileFormat';
import './PrefixFileFormat';
