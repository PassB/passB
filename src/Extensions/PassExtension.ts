import {PassCli} from "../PassCli";
import {Extension, ListEntry, RegisterEntryCallback} from "./Extension";

export class PassExtension extends Extension {
  public name = 'Base';
  public actions = ['show'];

  public async initializeList(registerEntryCallback: RegisterEntryCallback): Promise<void> {
    for (const label of await PassCli.list()) {
      registerEntryCallback({
        label,
        actions: this.actions,
        metadata: {},
      });
    }
    return void 0;
  }

  public executeAction(action: string, entry: ListEntry): void {
    switch (action) {
      case 'show':
        break;
      default:
        console.error('unknown action:', action);
    }
  }
}
