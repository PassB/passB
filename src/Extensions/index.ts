export {Extension, ExtensionTag, EntryActions, ExecutionOptions, RegisterEntryCallback} from './Extension';

import './PassExtension';
import './QRCodeExtension';
