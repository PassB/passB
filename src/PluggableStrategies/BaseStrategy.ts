import {Service} from 'typedi/build/compiled/src';
import {OptionsPanelType} from 'Options/OptionsReceiver';
import {getStrategyOptions, setStrategyDefaultOptions, OptionsDataType, StrategyName, StrategyType} from 'State/Options';
import {LazyInject} from '../Decorators/LazyInject';
import {State} from '../State/State';

@Service()
export abstract class BaseStrategy<OptionType> {
  public abstract readonly defaultOptions: OptionsDataType<OptionType>;
  public abstract readonly OptionsPanel?: OptionsPanelType<OptionType>;
  public abstract readonly type: StrategyType;
  public abstract readonly name: StrategyName;

  @LazyInject(() => State)
  private state: State;

  public constructor() {
    this.state.dispatch(setStrategyDefaultOptions({
      strategyType: this.type,
      strategyName: this.name,
      options: this.defaultOptions,
    }));
  }

  protected get options(): OptionsDataType<OptionType> {
    return getStrategyOptions(this.state.getOptions(), this.type, this.name) as OptionsDataType<OptionType>;
  }
}
