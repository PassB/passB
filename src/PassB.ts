import {InjectTagged, Service} from 'typedi';
import {executeInCorrectContext, AsynchronousCallableServiceFactory} from 'Decorators/ExecuteInContext';
import {LazyInject} from 'Decorators/LazyInject';
import {EntryActions, Extension, ExtensionTag} from 'Extensions';
import {FileFormat, FileFormatTag} from 'PluggableStrategies/FileFormats';
import {Filler, FillerTag} from 'PluggableStrategies/Fillers';
import {Matcher, MatcherTag} from 'PluggableStrategies/Matchers';
import {Options, OptionsData, OptionsList} from './Options/Options';
import {OptionsReceiverInterface} from './Options/OptionsReceiver';

export interface Action {
  extension: string;
  action: string;
}

export interface EntryNode {
  name: string;
  fullPath: string;
  actions: Action[];
  children: { [label: string]: EntryNode };
}

function buildEntryNode({name, fullPath, actions = [], children = {}}: {
  name: string;
  fullPath: string;
  actions?: Action[];
  children?: { [label: string]: EntryNode };
}): EntryNode {
  return {
    name,
    fullPath,
    actions,
    children,
  };
}

@Service({factory: AsynchronousCallableServiceFactory(PassB)})
export class PassB {
  @InjectTagged(ExtensionTag)
  protected extensions: Array<Extension<{}>>;
  @InjectTagged(FileFormatTag)
  protected fileFormats: Array<FileFormat<{}>>;
  @InjectTagged(MatcherTag)
  protected matchers: Array<Matcher<{}>>;
  @InjectTagged(FillerTag)
  protected fillers: Array<Filler<{}>>;
  @LazyInject(() => Options)
  protected options: Options;

  private rootNode: EntryNode = buildEntryNode({fullPath: '', name: ''});

  public async initialize(): Promise<this> {
    if (window.executionContext === 'background') {
      this.injectIntoChildren(await this.getOptions());
      await this.reloadEntries();
      return this;
    } else {
      this.injectIntoChildren(await this.getOptions());
      return this;
    }
  }

  public getAllExtensions(): Array<Extension<{}>> {
    return this.extensions;
  }

  public getExtension(name: string): Extension<{}> {
    const extension = this.extensions.find((item: Extension<{}>) => item.name === name);
    if (!extension) {
      throw new Error(`query for unknown extension ${name}`);
    }
    return extension;
  }

  public async getMatcher(): Promise<Matcher<{}>> {
    const selected = (await this.getOptions()).selectedMatcher;
    return this.matchers.find((strategy: OptionsReceiverInterface<{}>) => strategy.name === selected)
      || this.matchers[0];
  }

  public getAllMatchers(): Array<Matcher<{}>> {
    return this.matchers;
  }

  public async getFiller(): Promise<Filler<{}>> {
    const selected = (await this.getOptions()).selectedFiller;
    return this.fillers.find((strategy: OptionsReceiverInterface<{}>) => strategy.name === selected)
      || this.fillers[0];
  }

  public getAllFillers(): Array<Filler<{}>> {
    return this.fillers;
  }

  public async getFileFormat(): Promise<FileFormat<{}>> {
    const selected = (await this.getOptions()).selectedFileFormat;
    return this.fileFormats.find((strategy: OptionsReceiverInterface<{}>) => strategy.name === selected)
      || this.fileFormats[0];
  }

  public getAllFileFormats(): Array<FileFormat<{}>> {
    return this.fileFormats;
  }

  @executeInCorrectContext()
  @Reflect.metadata('executionContext', 'background')
  public async getRootNode(): Promise<EntryNode> {
    return this.rootNode;
  }

  @executeInCorrectContext()
  @Reflect.metadata('executionContext', 'background')
  public async getOptions(): Promise<OptionsData> {
    return await this.options.getOptions();
  }

  @executeInCorrectContext()
  @Reflect.metadata('executionContext', 'background')
  public async setOptions(newOptions: OptionsData): Promise<OptionsData> {
    await this.options.setOptions(newOptions);
    this.injectIntoChildren(newOptions);
    await this.reloadEntries();
    return newOptions;
  }

  @executeInCorrectContext()
  @Reflect.metadata('executionContext', 'background')
  private async reloadEntries(): Promise<this> {
    const enabledExtensionNames = (await (this.getOptions())).enabledExtensions;
    const enabledExtensions = this.extensions
      .filter((extension: Extension<{}>) => enabledExtensionNames.includes(extension.name));

    this.rootNode = buildEntryNode({fullPath: '', name: ''});
    const rootNode = this.rootNode; // keep reference to "current" rootNode to prevent potential timing issues

    const addEntries = (newActions: EntryActions, extension: Extension<{}>) => {
      const parts = newActions.label.split('/');
      let node: EntryNode = rootNode;
      for (const [depth, part] of Array.from(parts.entries())) {
        if (part !== '') {
          const partName = depth === parts.length - 1 ? part : `${part}/`;

          if (!node.children[partName]) {
            node.children[partName] = buildEntryNode({
              name: part,
              fullPath: node.fullPath + partName,
            });
          }
          node = node.children[partName];
        }

        if (depth === parts.length - 1) {
          node.actions.push(...newActions.actions.map((action: string) => ({extension: extension.name, action})));
        }
      }
    };

    for (const extension of enabledExtensions) {
      await (extension.initializeList((newActions: EntryActions) => addEntries(newActions, extension)));
    }

    return this;
  }

  private injectIntoChildren(options: OptionsData): void {
    const injectOptions = (optionsList: OptionsList) =>
      (item: OptionsReceiverInterface<{}>) => item.injectOptions(optionsList[item.name]);

    this.extensions.forEach(injectOptions(options.extensionsOptions));
    this.fileFormats.forEach(injectOptions(options.fileFormats));
    this.fillers.forEach(injectOptions(options.fillers));
    this.matchers.forEach(injectOptions(options.matchers));
  }
}
