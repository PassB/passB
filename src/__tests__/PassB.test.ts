// tslint:disable:no-var-requires max-classes-per-file
import {PassB as PassBType} from 'PassB';
import {Extension} from '../Extensions/Extension';
import {BaseStrategy} from '../PluggableStrategies/BaseStrategy';

/**
 * prevent side-effect imports
 * usually importing one of these would register all their implementations as tagged services
 */
function mockPreventSideEffects(): void {
  jest.setMock('Extensions', require('Extensions/Extension'));
  jest.setMock('PluggableStrategies/FileFormats', require('PluggableStrategies/FileFormats/FileFormat'));
  jest.setMock('PluggableStrategies/Fillers', require('PluggableStrategies/Fillers/Filler'));
  jest.setMock('PluggableStrategies/Matchers', require('PluggableStrategies/Matchers/Matcher'));
}

describe('PassB', () => {
  let {Container} = require('typedi');
  let {PassB} = require('PassB');
  let {setExecutionContext} = require('Decorators/ExecuteInContext');

  beforeEach(() => {
    jest.resetModules();
    setExecutionContext = require('Decorators/ExecuteInContext').setExecutionContext;
    Container = require('typedi').Container;
    PassB = require('PassB').PassB;
  });

  describe('tagged injections', () => {
    let passB: PassBType;
    beforeAll(() => {
      setExecutionContext('test1');
      passB = Container.get(PassB);
    });

    const mapToClassName = (item: BaseStrategy<{}> | Extension<{}>) => item.name;

    it('receives Extensions', () => {
      expect(passB.getAllExtensions().map(mapToClassName)).toMatchSnapshot();
    });

    it('receives FileFormats', () => {
      expect(passB.getAllFileFormats().map(mapToClassName)).toMatchSnapshot();
    });

    it('receives Fillers', () => {
      expect(passB.getAllFillers().map(mapToClassName)).toMatchSnapshot();
    });

    it('receives Matchers', () => {
      expect(passB.getAllMatchers().map(mapToClassName)).toMatchSnapshot();
    });
  });

  describe('extension behaviour', () => {
    beforeAll(mockPreventSideEffects);

    describe('extension injection', () => {

      it('receives extensions via the ExtensionTag', () => {
        const {Service} = require('typedi');
        const {ExtensionTag} = require('Extensions');

        @Service({tags: [ExtensionTag]})
        class Extension1 {
        }

        @Service({tags: [ExtensionTag]})
        class Extension2 {
        }

        setExecutionContext('background');
        const passB = Container.get(PassB);

        expect(passB.getAllExtensions().some((item: {}) => item instanceof Extension1)).toBeTruthy();
        expect(passB.getAllExtensions().some((item: {}) => item instanceof Extension2)).toBeTruthy();
        expect(passB.getAllExtensions().length).toBe(2);
      });

      it('receives FileFormats wia the FileFormatTag');
      it('receives Fillers wia the FillerTag');
      it('receives Matchers wia the MatcherTag');

    });

    it('validates that all extensions and strategies are of correct type at initialization');

    describe('injection of options', () => {
      //
    });
  });

  describe('context-specific behaviour', () => {
    it('injects options into children at initialize in any context');
    it('only starts loading entries in background context');
  });

  describe('loading on entries', () => {
    it('only triggers initializeList on enabled extensions');
    it('does not execute two extensions\' initializeList at the same time');
    it('waits for all extensions\' initializeList to finish');
    describe('builds a correct tree from the returned entries', () => {
      //
    });
  });
});
