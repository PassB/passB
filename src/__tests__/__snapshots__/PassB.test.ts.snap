// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PassB tagged injections receives Extensions 1`] = `
Array [
  "Pass",
  "QRCode",
]
`;

exports[`PassB tagged injections receives FileFormats 1`] = `
Array [
  "FirstLineFileFormat",
  "PrefixFileFormat",
]
`;

exports[`PassB tagged injections receives Fillers 1`] = `
Array [
  "FillPasswordInputs",
]
`;

exports[`PassB tagged injections receives Matchers 1`] = `
Array [
  "FuzzaldrinMatcher",
]
`;
